### PR TITLE
Refactoring to drop extensions dir, see #774

### PR DESCRIPTION
**Pull request by Luceos**
- See #774 & https://github.com/flarum/flarum/pull/35
- No longer loading the composer.json in the extension directory, but fully utilizing the installed.json of composer.

### tested:

- with dev-master of all core extensions and dev-master of flarum/flarum dropping flarum/composer-installer
- loading flarum
- loading admin
- en-/disabling extensions
- re-running migrations

### CODE REVIEW

**Comment by Luceos** [comment]
Is styleci complaining about a non-existent tab? :laughing:

**Comment by tobscure** [comment]
Great work! Will test as soon as I get time. 

StyleCI is complaining about extra unnecessary whitespace on that line.

File: src/Extension/Extension.php [hunk 1]
```diff
@@ -96,8 +96,9 @@ public function __construct($path, $composerJson)
      */
     protected function assignId()
     {
-        $segments = explode('/', $this->path);
-        $this->id = end($segments);
+        list($vendor, $package) = explode('/', $this->name);
+        $package = str_replace(['flarum-ext-', 'flarum-'], null, $package);
```
Comment (franzliedke) [inline src/Extension/Extension.php]: Hmm, minor thing: It surprised me that the second parameter to `str_replace` can be `null`. Can you change it to an empty string? That's really what we want to express here, right?

File: src/Extension/ExtensionManager.php [hunk 2]
```diff
@@ -260,13 +282,24 @@ protected function setEnabled(array $enabled)
...
 
+    /**
+     * The extensions path.
+     *
+     * @return string
+     */
     protected function getExtensionsDir()
     {
-        return public_path('extensions');
+        return public_path('vendor');
```
Comment (franzliedke) [inline src/Extension/ExtensionManager.php]: Can we pass that in when defining the manager in the service provider?

File: src/Extension/ExtensionServiceProvider.php [hunk 3]
```diff
@@ -21,11 +21,10 @@ public function register()
     {
         $this->app->bind('flarum.extensions', 'Flarum\Extension\ExtensionManager');
 
-        $config = $this->app->make('flarum.settings')->get('extensions_enabled');
-        $extensions = json_decode($config, true);
+        $extensions = $this->app->make('flarum.extensions')->getEnabledExtensions();
 
         foreach ($extensions as $extension) {
-            if (file_exists($file = public_path().'/extensions/'.$extension.'/bootstrap.php')) {
+            if (file_exists($file = $extension->getPath().'/bootstrap.php')) {
```
Comment (franzliedke) [inline src/Extension/ExtensionServiceProvider.php]: Can we shorten this to a method in the extension manager that returns an array of files to be included?

**Comment by franzliedke** [comment]
Looks good, overall! :+1: 
Just some minor corrections...

**Comment by franzliedke** [comment]
Oh, last thing: do we still get any benefit from defining a special type for extension composer packages? Or was that only needed for installing it to a different directory?

**Comment by tobscure** [comment]
@franzliedke we need the flarum-extension type to be able to tell which packages in installed.json are Flarum extensions :)

File: src/Extension/ExtensionManager.php (hunk 2 above)
Comment (Luceos) [inline src/Extension/ExtensionManager.php]: While loading the installed.json we also use the hardcoded public path to vendor. IMO that's wrong too. So I just kept it like this. We might need to provide a way to get the vendor_path and extension_path seperately from the ExtensionManager, perhaps in the Application class. Same behavior as Laravel in that sense.

If you don't mind, won't be taking this along for this PR.

**Comment by Luceos** [comment]
@flarum/core 

> Just some minor corrections...

All done except for the one to pass the vendor folder into the construct of the Extension Manager, I think there should be an alternative solution for that. Either create a `extension_path` and `vendor_path` helper or implement those paths in the Application.

**Comment by tobscure** [comment]
Fix up the StyleCI error and I'm happy :) @franzliedke we can work out a solution for the hardcoded vendor path after

**Comment by Luceos** [comment]
@tobscure was looking at those.. I really need to set up phpstorm to use the styleci/nitpick params.

**Comment by Luceos** [comment]
@tobscure done.

**Comment by tobscure** [comment]
Thanks so much @Luceos, great work :)

**Comment by Luceos** [comment]
Happy with this too :)

Don't forget about https://github.com/flarum/flarum/pull/35
